### PR TITLE
fix the case of time drift and re-implement next schedule calculation

### PR DESCRIPTION
**Pull request by alaypatel07**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

This PR fixes a bug that could create a panic if time drifts back. The test case for time drifting back is refined to catch the error. It was found that if the job is not created, the `jobResp` will be nil, however, it is used as if the job was actually created. This PR gets the job from the job lister if create call fails and populates the `jobResp` making it non-nil and safe for use further in the reconcile loop

This PR also re-implements the way the next schedule is calculated. Instead of iterating over the next schedules, this PR calculates the schedule which is an optimization over previous version.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/cronjob/utils.go [hunk 1]
```diff
@@ -150,12 +150,12 @@ func getRecentUnmetScheduleTimes(cj batchv1beta1.CronJob, now time.Time) ([]time
 	return starts, nil
 }
 
-// getUnmetScheduleTimes gets the slice of all the missed times from the time a job
+// getNextScheduleTime gets the slice of all the missed times from the time a job
```
Comment (soltysh) [inline pkg/controller/cronjob/utils.go:153]: Nit: the doc says about slice of times, but that's not true any more, this returns a single time time pointer now.

File: pkg/controller/cronjob/utils_test.go [hunk 2]
```diff
@@ -426,8 +421,8 @@ func TestGetLatestUnmetScheduleTimes(t *testing.T) {
 		// Deadline is short
 		deadline := int64(2 * 60 * 60)
 		cj.Spec.StartingDeadlineSeconds = &deadline
-		times := getUnmetScheduleTimes(cj, now, PraseSchedule(cj.Spec.Schedule), recorder)
-		if len(times) == 0 {
+		schedule := getNextScheduleTime(cj, now, PraseSchedule(cj.Spec.Schedule), recorder)
+		if schedule == nil {
```
Comment (soltysh) [inline pkg/controller/cronjob/utils_test.go:425]: These entire test case are begging to be re-written as table tests.

File: pkg/controller/cronjob/utils_test.go [hunk 3]
```diff
@@ -631,3 +626,65 @@ func TestByJobStartTime(t *testing.T) {
 		}
 	}
 }
+
+func Test_getMostRecentScheduleTime(t *testing.T) {
```
Comment (soltysh) [inline pkg/controller/cronjob/utils_test.go:630]: `TestGetMostRecentScheduleTime`

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 4]
```diff
@@ -556,13 +555,20 @@ func (jm *ControllerV2) syncCronJob(
 	case errors.HasStatusCause(err, corev1.NamespaceTerminatingCause):
 	case errors.IsAlreadyExists(err):
 		klog.InfoS("Job already exists", "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()), "job", klog.KRef(jobReq.GetNamespace(), jobReq.GetName()))
+		var getError error
+		jobResp, getError = jm.jobControl.GetJob(jobReq.Namespace, jobReq.Name)
+		if getError != nil {
+			return cj, nil, getError
+		}
```
Comment (soltysh) [inline pkg/controller/cronjob/cronjob_controllerv2.go:562]: Don't do that, if a job was created by a different actor (other worker) then it'll also have taken care of updating status and we shouldn't do anything with it. Just continue without taking any actions.

File: pkg/controller/cronjob/cronjob_controllerv2.go [hunk 5]
```diff
@@ -556,13 +555,20 @@ func (jm *ControllerV2) syncCronJob(
...
 		// default error handling
 		jm.recorder.Eventf(cj, corev1.EventTypeWarning, "FailedCreate", "Error creating job: %v", err)
 		return cj, nil, err
+	default:
+		// error is nil, job was just created
+		klog.V(4).InfoS("Created Job", "job", klog.KRef(jobResp.GetNamespace(), jobResp.GetName()), "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()))
+		jm.recorder.Eventf(cj, corev1.EventTypeNormal, "SuccessfulCreate", "Created job %v", jobResp.Name)
 	}
-	klog.V(4).InfoS("Created Job", "job", klog.KRef(jobResp.GetNamespace(), jobResp.GetName()), "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()))
-	jm.recorder.Eventf(cj, corev1.EventTypeNormal, "SuccessfulCreate", "Created job %v", jobResp.Name)
```
Comment (soltysh) [inline pkg/controller/cronjob/cronjob_controllerv2.go:565]: Based on my previous comment, that should not be needed.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: pkg/controller/cronjob/utils.go (hunk 1 above)
Comment (alaypatel07) [reply pkg/controller/cronjob/utils.go]: updated, PTAL

File: pkg/controller/cronjob/utils_test.go (hunk 2 above)
Comment (alaypatel07) [reply pkg/controller/cronjob/utils_test.go:425]: +1, will try to take it as a follow-up PR

File: pkg/controller/cronjob/utils_test.go (hunk 3 above)
Comment (alaypatel07) [reply pkg/controller/cronjob/utils_test.go]: updated.

File: pkg/controller/cronjob/cronjob_controllerv2.go (hunk 4 above)
Comment (alaypatel07) [reply pkg/controller/cronjob/cronjob_controllerv2.go]: +1, removed

File: pkg/controller/cronjob/cronjob_controllerv2.go (hunk 5 above)
Comment (alaypatel07) [reply pkg/controller/cronjob/cronjob_controllerv2.go:565]: +1 dropped the change

**Comment by alaypatel07** [comment]
retrying, the failure seemed unrelated to the PR

/test pull-kubernetes-e2e-gce-ubuntu-containerd

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
/release-note none
/triage accepted
/priority backlog

**Comment by liggitt** [comment]
flaky cronjob tests (https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&text=CronJob&job=unit%7Cbazel) coincide with this merging

**Pull request merged**